### PR TITLE
[RFC] Trusted Service abstraction

### PR DESCRIPTION
**Pull request by jenswi-linaro**
As a step in making room for Secure Partitions (SPs) running at S-EL0 add a Trusted Service (TS) abstraction. Both TAs and SPs is a TS.

The idea is to form a struct ts_session and struct ts_ctx and so on that can be used by common code that might deal with both TAs and SPs.

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
I'm working on the various compile errors.

**Review by jellesels-arm — COMMENTED** [review]
- I am wondering if we she move sbuf and fbuf from the tee_ta_sessions to the ts_session. This would make it easier to implement tracing for all trusted services.

**Review by jellesels-arm — APPROVED** [review]
Looks good to me

**Review by jellesels-arm — COMMENTED** [review]
Looks good

**Review by jellesels-arm — COMMENTED** [review]
Looks good

File: core/kernel/tee_ta_manager.c [hunk 1]
```diff
@@ -821,79 +824,16 @@ bool tee_ta_session_is_cancelled(struct tee_ta_session *s, TEE_Time *curr_time)
...
+	struct ts_session *s = ts_get_current_session();
 	struct user_ta_ctx *utc = NULL;
 	struct sample_buf *sbuf = NULL;
 	TEE_Result res = 0;
 	size_t idx = 0;
 
-	if (tee_ta_get_current_session(&s) != TEE_SUCCESS)
-		return;
-	sbuf = s->sbuf;
+	sbuf = to_ta_session(s)->sbuf;
```
Comment (etienne-lms) [inline core/kernel/tee_ta_manager.c:850]: Previous implementation safely returned without segfault when no current session.
Looks ok here, but maybe cross check this is safe.

File: core/tee/socket.c [hunk 2]
```diff
@@ -225,11 +225,10 @@ static TEE_Result pta_socket_open_session(uint32_t param_types __unused,
 			TEE_Param pParams[TEE_NUM_PARAMS] __unused,
 			void **sess_ctx)
 {
-	struct tee_ta_session *s;
+	struct ts_session *s = ts_get_calling_session();
 
 	/* Check that we're called from a TA */
-	s = tee_ta_get_calling_session();
-	if (!s)
+	if (!is_user_ta_ctx(s->ctx))
```
Comment (etienne-lms) [inline core/tee/socket.c:231]: If this PTA is called from REE, `ts_get_calling_session()` will return `NULL` hence a segfault here.
Maybe replace with `if (!s || !is_user_ta_ctx(s->ctx))`.

**Review by etienne-lms — COMMENTED** [review]
Review for commit "core: add generic struct ts_session".

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits
"core: replace TA with user mode in abort handling"
"core: add generic struct ts_ctx"
"core: add is_ta_ctx()"
"core: add struct ts_ops"
"core: store TA params in session struct"

**Comment by jellesels-arm** [comment]
Can we use a pointer to a struct ts_ctx  instead of the struct tee_ta_ctx  in user_mode_ctx?
This will make it easier to integrate the SP version of the user_ta_ctx struct.

**Comment by jenswi-linaro** [comment]
> * I am wondering if we she move sbuf and fbuf from the tee_ta_sessions to the ts_session. This would make it easier to implement tracing for all trusted services.

Makes sense, I'll fix that.

**Comment by jenswi-linaro** [comment]
> Can we use a pointer to a struct ts_ctx instead of the struct tee_ta_ctx in user_mode_ctx?
> This will make it easier to integrate the SP version of the user_ta_ctx struct.

I'll try to think of something.

**Comment by jenswi-linaro** [comment]
Update.
We may want to squash "core: move sbuf and fbuf into struct ts_session" into "core: add generic struct ts_session".
"core: remove struct tee_ta_ctx from struct user_mode_ctx" probably makes sense to keep as a separate commit.

**Comment by jenswi-linaro** [comment]
Rebased to resolve the merge conflict

**Comment by jenswi-linaro** [comment]
Rebased and master. Had to resolve a few tricky merge conflicts so I haven't applied the tags.
Folded  "core: move sbuf and fbuf into struct ts_session" into "core: add generic struct ts_session".

File: core/pta/system.c [hunk 3]
```diff
@@ -40,9 +40,9 @@ struct system_ctx {
...
 
-static TEE_Result system_rng_reseed(struct tee_ta_session *s __unused,
-				uint32_t param_types,
-				TEE_Param params[TEE_NUM_PARAMS])
+static TEE_Result system_rng_reseed(struct ts_session *s __unused,
+				    uint32_t param_types,
+				    TEE_Param params[TEE_NUM_PARAMS])
 {
 	size_t entropy_sz;
 	uint8_t *entropy_input;
```
Comment (etienne-lms) [inline core/pta/system.c:48]: minor:could you add some init values here?

File: core/tee/socket.c (hunk 2 above)
Comment (etienne-lms) [reply core/tee/socket.c:231]: comment still applies (I think).

**Review by etienne-lms — COMMENTED** [review]
For commit "core: add generic struct ts_session".

**Review by etienne-lms — COMMENTED** [review]
For commit "core: store TA params in session struct":
typo in commit log: s/passing it/passing them/

For commit "core: remove struct tee_ta_ctx from struct user_mode_ctx":
typo in commit log: s/function/functions/

LGTM for the whole series aside the few comments I posted.

File: core/arch/arm/kernel/pseudo_ta.c [hunk 4]
```diff
@@ -306,7 +308,7 @@ TEE_Result tee_ta_init_pseudo_ta_session(const TEE_UUID *uuid,
 	ctx->ops = &pseudo_ta_ops;
 
 	mutex_lock(&tee_ta_mutex);
-	s->ctx = ctx;
+	s->ts_sess.ctx = ctx;;
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/kernel/pseudo_ta.c]: I'll move the fix into this commit.

File: core/pta/system.c (hunk 3 above)
Comment (jenswi-linaro) [reply core/pta/system.c:48]: OK

File: core/pta/system.c [hunk 5]
```diff
@@ -844,11 +844,11 @@ static TEE_Result open_session(uint32_t param_types __unused,
 			       TEE_Param params[TEE_NUM_PARAMS] __unused,
 			       void **sess_ctx)
 {
-	struct tee_ta_session *s = NULL;
+	struct ts_session *s = NULL;
 	struct system_ctx *ctx = NULL;
 
 	/* Check that we're called from a user TA */
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/pta/system.c:850]: For now I think `is_user_ta_ctx()` is good enough.

File: core/tee/socket.c (hunk 2 above)
Comment (jenswi-linaro) [reply core/tee/socket.c:231]: I'm sorry, I missed this. I'll fix.

File: core/include/kernel/tee_ta_manager.h [hunk 6]
```diff
@@ -183,4 +177,10 @@ static inline void tee_ta_ftrace_update_times_suspend(void) {}
 static inline void tee_ta_ftrace_update_times_resume(void) {}
 #endif
 
+static inline struct tee_ta_session * __noprof
+to_ta_session(struct ts_session *sess)
+{
+	return container_of(sess, struct tee_ta_session, ts_sess);
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/include/kernel/tee_ta_manager.h:167]: I'm not sure that's possible to check at the point of the patch, but there's a test added later in this PR.

**Comment by jenswi-linaro** [comment]
Comments addressed and tag applied, assuming LGTM meant R-B.

**Review by jforissier — COMMENTED** [review]
Commit "core: add generic struct ts_session" does not say what "ts" means and why it is better than the current naming. I think the explanation text in the first comment https://github.com/OP-TEE/optee_os/pull/4126#issue-498584829 would fit quite well.

**Comment by jenswi-linaro** [comment]
Updated the commit message in "core: add generic struct ts_session".

**Comment by jforissier** [comment]
Would you mind fixing the checkpatch warning? Then I will merge.

**Comment by jenswi-linaro** [comment]
> Would you mind fixing the checkpatch warning? Then I will merge.

I'm sorry, I had missed that. Should be OK now.

**Comment by jenswi-linaro** [comment]
Now there was another warning, fixing that too.

**Pull request merged**